### DISCUSSION TITLE
Websocket 2-way connection

### DISCUSSION THREAD

**Discussion by okaforcj** [trpc/trpc#2242] [Q&A] [answered]
I know websockets are still in beta but I've got a couple questions to further my understanding. 

First off, the websocket implementation doesn't seem like true 2-way connection or I might be missing something, it seems like its just doing a HTTP SSE (flat subscription) in my opinion, I thought the idea of websockets where true 2 way connections i.e. the client can actually message the server. This is particularly painful because even though an expensive websocket connection is open, I'm having to take up more memory and bandwidth (potentially) to open up a redundant TCP connection to send a message to the server. Also, because my users are authenticated I can't find an immediate way to associate a userId to a socketId. This would have been possible in a 2-way connection because when the message is received on the server from the client, the socketid that sent it can already be known. My eventual question is please what are the workarounds cos I was using socket.io to do all these things with no fuss, everything else on my app (HTTP requests) was typed with trpc so I figured to extend to subscriptions

**Comment by KATT** [comment, marked as answer]
I really don't understand what you're asking for you or how the design is wrong.

We have a real-time chat example with authentication where all communication happens over websockets: https://github.com/trpc/examples-next-prisma-starter-websockets

**Reply by okaforcj** [reply]
thanks for the reply. and woahh i didn't say there was anything wrong with the design, like I said I am just looking for more clarity. Also, I went through that example before I raised this. Their example doesn't send anything through the websocket from client to server.
For example, to add a post, you have to add a mutation (which they did). I would have assumed that rather than sending a mutation you could emit data over the websocket connection from the front-end? Could you clarify how one could emit data from client to server ? Also I wasn't saying  you cant authenticate with it cos I am currently chatting with users on my app in an authenticated state, I was more asking how you could find out the information of a user from their socketId. I am saying this because I have a server run on google cloud run which is stateless and so I need to know which user is sending information from a particular socketId at any given time. I hope my questions are better clarified @KATT ?

**Reply by KATT** [reply]
The queries & mutations are sent over WebSockets using the same connection as the subscriptions, please clone the project and look at the network tab.

**Reply by henninghall** [reply]
@KATT Would mutations sent through websockets require a response every time? 

In that case it would still be adding some overhead compared to regular websockets where we can have high frequency streaming (client->server) without requiring a response right? Or am I getting something wrong?

**Reply by KATT** [reply]
Sure, a response is always sent, but you don't need to `await` it, you can do fire-and-forget
